File: VIF_git/GSDS-Qual/algorithms/Kmeans/lloyds.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <limits>
#include <cstdlib>
#include <ctime>

using namespace std;

struct Point {
    double x, y;
    int cluster; // Cluster id

    Point(double a, double b) : x(a), y(b), cluster(0) {};
    Point(double a, double b, int c) : x(a), y(b), cluster(c) {};
};

double euclideanDistance(const Point& p1, const Point& p2) {
    return sqrt((p1.x - p2.x) * (p1.x - p2.x) + (p1.y - p2.y) * (p1.y - p2.y));
}

void initializeCentroids(vector<Point>& centroids, const vector<Point>& points, int k) {
    srand(time(0));                                      // 초기화를 위함
    for (int i = 0; i < k; ++i) {
        centroids[i] = points[rand() % points.size()];   // 랜덤한 point로 모든 centroid를 배정
    }
}

// 한 point와 centroid들이 주어졌을 때 그 point와 가장 가까운 centroid를 골라주는 함수
int findClosestCentroid(const Point& point, const vector<Point>& centroids) {
    int bestIndex = 0;
    double bestDist = numeric_limits<double>::max();

    for (int i = 0; i < centroids.size(); ++i) {
        double dist = euclideanDistance(point, centroids[i]);
        if (dist < bestDist) {
            bestDist = dist;
            bestIndex = i;
        }
    }
    return bestIndex;
}

void updateCentroids(vector<Point>& centroids, const vector<Point>& points, int k) {
    vector<int> count(k, 0);
    vector<Point> newCentroids;
    for (int i = 0; i < k; i++) newCentroids.push_back(Point(0, 0, 0));

    for (const auto& point : points) {
        newCentroids[point.cluster].x += point.x;
        newCentroids[point.cluster].y += point.y;
        count[point.cluster]++;
    } // 각각의 cluster별로 해당하는 point들의 값을 더해줌 (나중에 평균 낼 용도)

    for (int i = 0; i < k; ++i) {
        if (count[i] > 0) {
            centroids[i].x = newCentroids[i].x / count[i];
            centroids[i].y = newCentroids[i].y / count[i];
        }
    }
}

bool hasConverged(const vector<Point>& oldCentroids, const vector<Point>& newCentroids) {
    for (int i = 0; i < oldCentroids.size(); ++i) {
        if (euclideanDistance(oldCentroids[i], newCentroids[i]) > 1e-6) {
            return false;
        } // 근소한 정도 (1e-6) 이상의 차이가 난다면 converge하지 않았다 판단
    }
    return true;
}

void kMeans(vector<Point>& points, int k) {
    vector<Point> centroids;
    for (int i = 0; i < k; i++) centroids.push_back(Point(0, 0, 0));
    initializeCentroids(centroids, points, k);

    vector<Point> oldCentroids;
    bool converged = false;

    while (!converged) {
        oldCentroids = centroids;

        for (auto& point : points) {
            point.cluster = findClosestCentroid(point, centroids);
        } // 모든 point들마다 가장 가까운 centroid 찾아줌

        updateCentroids(centroids, points, k);
        converged = hasConverged(oldCentroids, centroids);
    }
}

int main() {
    vector<Point> points;
    points.push_back(Point(1.0, 2.0, 0));
    points.push_back(Point(1.5, 1.8, 0));
    points.push_back(Point(5.0, 8.0, 0));
    points.push_back(Point(8.0, 8.0, 0));
    points.push_back(Point(1.0, 0.6, 0));
    points.push_back(Point(9.0, 11.0, 0));
    points.push_back(Point(8.0, 2.0, 0));
    points.push_back(Point(10.0, 2.0, 0));
    points.push_back(Point(9.0, 3.0, 0));


    int k = 3;
    kMeans(points, k);

    for (const auto& point : points) {
        cout << "Point (" << point.x << ", " << point.y << "): Cluster " << point.cluster << endl;
    }

    return 0;
}

File: VIF_git/GSDS-Qual/algorithms/Kmeans/lloyds.py
import random
import math
import sys

class Point:
    def __init__(self, x, y, cluster=0):
        self.x = x
        self.y = y
        self.cluster = cluster

# 구조는 다른 언어들과 같게 써봤는데 실전에선 이런 식으로 함수 정의 안 해도 될 것 같애요...

def euclidean_distance(p1, p2):
    return math.sqrt((p1.x - p2.x) ** 2 + (p1.y - p2.y) ** 2)

def initialize_centroids(points, k):
    centroids = random.sample(points, k)
    return centroids

def find_closest_centroid(point, centroids):
    best_index = 0
    best_dist = sys.float_info.max

    for i, centroid in enumerate(centroids):
        dist = euclidean_distance(point, centroid)
        if dist < best_dist:
            best_dist = dist
            best_index = i
    
    return best_index

def update_centroids(points, centroids, k):
    new_centroids = [Point(0, 0) for _ in range(k)]
    count = [0] * k

    for point in points:
        new_centroids[point.cluster].x += point.x
        new_centroids[point.cluster].y += point.y
        count[point.cluster] += 1
    
    for i in range(k):
        if count[i] > 0:
            new_centroids[i].x /= count[i]
            new_centroids[i].y /= count[i]
    
    return new_centroids

def has_converged(old_centroids, new_centroids):
    for old, new in zip(old_centroids, new_centroids):
        if euclidean_distance(old, new) > 1e-6:
            return False
    return True

def k_means(points, k):
    centroids = initialize_centroids(points, k)
    converged = False

    while not converged:
        for point in points:
            point.cluster = find_closest_centroid(point, centroids)
        
        new_centroids = update_centroids(points, centroids, k)
        converged = has_converged(centroids, new_centroids)
        centroids = new_centroids
    
    return centroids

if __name__ == "__main__":
    points = [
        Point(1.0, 2.0),
        Point(1.5, 1.8),
        Point(5.0, 8.0),
        Point(8.0, 8.0),
        Point(1.0, 0.6),
        Point(9.0, 11.0),
        Point(8.0, 2.0),
        Point(10.0, 2.0),
        Point(9.0, 3.0),
    ]

    k = 3
    k_means(points, k)

    for point in points:
        print(f"Point ({point.x}, {point.y}): Cluster {point.cluster}")
